File: leetcode/implementation/0289_game_of_life.py
# LC 289. Game of Life | Medium
class Solution:
    def gameOfLife(self, board):
        m, n = len(board), len(board[0])
        for i in range(m):
            for j in range(n):
                live = 0
                for di in (-1, 0, 1):
                    for dj in (-1, 0, 1):
                        if di == 0 and dj == 0: continue
                        ni, nj = i+di, j+dj
                        if 0 <= ni < m and 0 <= nj < n and board[ni][nj] in (1, 2):
                            live += 1
                if board[i][j] == 1 and (live < 2 or live > 3): board[i][j] = 2
                elif board[i][j] == 0 and live == 3: board[i][j] = 3
        for i in range(m):
            for j in range(n): board[i][j] %= 2

File: leetcode/implementation/0289_game_of_life.cpp
// LC 289. Game of Life | Medium
#include <vector>
using namespace std;

class Solution {
public:
    void gameOfLife(vector<vector<int>>& board) {
        int m = board.size(), n = board[0].size();
        for (int i = 0; i < m; i++)
            for (int j = 0; j < n; j++) {
                int live = 0;
                for (int di = -1; di <= 1; di++)
                    for (int dj = -1; dj <= 1; dj++) {
                        if (!di && !dj) continue;
                        int ni = i+di, nj = j+dj;
                        if (ni >= 0 && ni < m && nj >= 0 && nj < n && (board[ni][nj] & 1)) live++;
                    }
                if ((board[i][j] & 1) && (live == 2 || live == 3)) board[i][j] |= 2;
                else if (!(board[i][j] & 1) && live == 3) board[i][j] |= 2;
            }
        for (auto& row : board) for (auto& c : row) c >>= 1;
    }
};
